Correct Use of SslProtocols Enum

Kevin-Andrew opened the pull request:
To address feedback given in #47, this is a breaking change for anyone using the `SyslogTcpConfig` or `SyslogLoggerConfigurationExtensions.TcpSyslog()` extension method. The `SecureProtocols` property/parameter has been replace with a simple boolean, `UseTls`. If you were previously setting the `SecureProtocols` property/parameter to a value such as `SslProtocols.Tls12`, then you would set the new `UseTls` property/parameter to `true`. If you were using `SslProtocols.None`, then you will pass in `false` for the new boolean value.

Also updated in the PR are the certificate based unit tests to cleanup/delete the private keys after they are used. Otherwise, they become orphaned on the hard drive. (That is for Windows specifically, not sure what happens on Linux.)

Lastly, the main project, Serilog.Sinks.Syslog, is set to target just the .NET Standard 2.0 framework. Because we do not have any framework specific code, this one build will allow it to be run under any .NET, on any platform, that supports .NET Standard 2.0. This include .NET Framework 4.6.2+, .NET Core2.0+, and .NET 5+, on Windows, Linux, or MacOS.

Unit test results in Visual Studio on Windows:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/4dbaf22c-227c-4e5c-bd67-7184996deb6d)

Unit test results in VS Code on Ubuntu:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/a76ded45-28c6-4ef8-a009-a900ea4d3191)

@colin-anderson , I hope the changes to the project file(s) are compatible with the build system on GitHub. If not, we can revert it by deleting the new `Directory.Build.props` file and duplicating the items in each individual project file again.

Kevin-Andrew commented:
@colin-anderson , I see that AppVeyor is still unable to build .NET 6.0. I see in the .yml file that it is using a Visual Studio 2019 image, which cannot, by default build .NET 6.0. Do you know if that can be updated to use a Visual Studio 2022 image?

colin-anderson commented on src/Serilog.Sinks.Syslog/Sinks/SyslogTcpSink.cs:
```diff
@@ -233,6 +231,22 @@ private async Task<Stream> GetStream(Stream baseStream)
+                    // ErrorCode of ERROR_OPERATION_ABORTED is thrown as opposed to the ObjectDisposedException
+                    // above. We want to catch and ignore this error as well as it means the timeout value
+                    // has elapsed.
+                    var sex = ex.InnerException as SocketException;
```
`sex` -> `socketEx`? 😅

Kevin-Andrew replied:
Changed.

colin-anderson commented on test/Serilog.Sinks.Syslog.Tests/CertificateProviders/CertificateStoreProviderTests.cs:
```diff
@@ -4,20 +4,64 @@
 {
     public class CertificateStoreProviderTests : IDisposable
     {
+        // In Visual Studio, unit test can/do run in parallel, even for the different target frameworks. But these
```
Huh, have you actually had this happen in practice? I just had Rider execute these tests repeatedly, across all frameworks, for 10 minutes, without any failures.

Kevin-Andrew replied:
Are you asking just about the parallel execution, or about the failures? For the parallel execution, with Visual Studio, here is a screen shot before running the tests:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/55b31626-397e-45a3-96b8-23fbe92498ca)

![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/b0f0b029-a2f0-4c35-a0e4-900715b62f02)

As for the failures, if you are running the old test, where it didn't care/try to delete/clean up the private key file, then you are correct, there would be no failures, regardless of parallel execution or not.

However, by cleaning up the private key files, because it is a shared resource, if the tests execute in parallel, they can fail. If I simply remove the mutex, and rerun the tests, I will get a failure on at least one framework type every time.
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/dfe357e8-4fce-4d33-a12e-1718cc3a3f16)

colin-anderson commented on test/Serilog.Sinks.Syslog.Tests/CertificateProviders/CertificateStoreProviderTests.cs:
```diff
@@ -40,7 +84,105 @@ public void Dispose()
+        /// <param name="cert">The certificate that was loaded with a persisted key.</param>
+        /// <remarks>Only attempts to delete from the UserKeyset, which on Windows will be under the following:
+        /// C:\Users\[user]\AppData\Roaming\Microsoft\Crypto\RSA\S-1-5-21-....</remarks>
+        private static void DeletePrivateKey(X509Certificate2 cert)
```
Hmm, I had assumed that `X509Store.Remove(X509Certificate)` also removed the private key file from disk?! That's... very annoying if not!

Kevin-Andrew replied:
Unfortunately no. Microsoft has never done that, ever, in any operating system.

colin-anderson commented on src/Serilog.Sinks.Syslog/Sinks/Certificates/CertificateFileProvider.cs:
```diff
@@ -22,7 +22,11 @@ public CertificateFileProvider(string filename, string password = null)
                 throw new FileNotFoundException($"Certificate {filename} could not be found");
 
-            this.Certificate = new X509Certificate2(filename, password, X509KeyStorageFlags.PersistKeySet);
+            // Do not persist the private key. We do not have any code here to delete it. So the file ends up
```
Nice catch, this problem always annoyed me on my rare visits to the folder in question 😄

colin-anderson commented:
@Kevin-Andrew yes, it's as simple as changing `Visual Studio 2019` to `Visual Studio 2022` in the `appveyor.yml` file 😃 

Otherwise, all looks good to me!

Kevin-Andrew commented:
Trying `Visual Studio 2022` in the `appveyor.yml` file.

Kevin-Andrew commented:
Well, with `Visual Studio 2022` in the `appveyor.yml` file, we now get the .NET 6.0 unit tests to execute:
https://ci.appveyor.com/project/ionx-solutions/serilog-sinks-syslog/builds/47389842#L197
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/a9403216-f872-433a-85f5-5829ed712776)

And all the tests pass for that framework, but now the .NET Framework 4.6.2 test suite is showing an error. Didn't have that problem yesterday. It's unfortunate that I don't get the error on my machine to be able to easily debug it. Let me see what I can do.

Kevin-Andrew commented:
For the unit test error with the .NET Framework 4.6.2, I found what I believe is the reason here in this documentation:
https://learn.microsoft.com/en-us/dotnet/framework/network-programming/tls
specifically:
https://learn.microsoft.com/en-us/dotnet/framework/network-programming/tls#for-net-framework-46---462-and-not-wcf

But after trying to set that switch in both code and the app.config file, the error still occurs when running under AppVeyor.

I can try to do more testing and see if I can reproduce the problem on another machine.

Kevin-Andrew commented:
@colin-anderson , I was able to reproduce the `SslProtocols` error that AppVeyor reports when running the unit tests for .NET 4.6.2 on Windows Server 2019:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/f900a72b-af86-47a9-a137-18874eafee4b)

But I am also able to set the `AppContext` switch, as documented, in code, and have the application run successfully:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/b96e1d78-6d71-4f95-9b34-783af12ac2f4)

So I don't know why the same thing isn't working when running as a unit test under AppVeyor. I see a few options:
1. Ignore it and don't care.
2. Figure out a way to skip that test for just .NET 4.6.2.
3. Completely drop support for .NET 4.6.2 and up it to 4.7+.

Up to you. Let me know what you want to do.

Also, I was wrong about only needing .NET Standard 2.0. While it does work, I'm sure nobody wants all the extra assemblies being pulled into their output if they aren't targeting .NET Standard 2.0 themselves. So I changed it back and also added .NET 7.0.

Kevin-Andrew commented:
Performed another test. I wanted to see if there was a difference between the behavior of a standalone .NET 4.6.2 executable application, versus whatever the actual unit tests were being run as/under.

So I installed the .NET Core 3.1 SDK and the .NET 6.0 SDK on my Windows Server 2019 machine, after having applied all Windows updates. I could then run the `build.cmd` file from the serilog-sinks-syslog folder, just like, what I believe, AppVeyor is doing.

And again, I am able to both reproduce the problem (by commenting out the code fix that is currently there), as well as have the unit test execute successfully.

Here is the unit tests executing successfully for .NET 4.6.2. The unit test code has the code to set the `AppContext` switch according to the documentation.
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/84b91e85-82c4-4528-9e24-128060b599c1)

And if I comment out the code to set the `AppContext` switch, the unit test will fail, just like it does when running in AppVeyor:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/be3ec04f-70c8-41aa-9e77-dd6948a918df)

So there must still be some difference about the AppVeyor image compared to what I am testing with. I could try to make yet another Windows Server 2019 image with perhaps a newer build number. Currently, mine is build 1809. I do not know what build AppVeyor is running.

antibarbie commented:
@Kevin-Andrew about the 4.6.x failures, I think the microsoft documentation on cipher suites and TLS versions says that the behaviour changed with 4.7.x ; the best practices are documented here https://learn.microsoft.com/en-us/dotnet/framework/network-programming/tls and things looks more confused with v4.6 or v3.5 ; I understand that the "automatic mode" with "None" is only for v4.7+

Kevin-Andrew commented:
@antibarbie , yup. I included that same link [three comments above](https://github.com/IonxSolutions/serilog-sinks-syslog/pull/63#issuecomment-1606258670). Following all of that documentation, even looking at the Microsoft .NET Framework source code, I can see where, why, and how the setting is being applied. And I get expected behavior for both positive and negative tests when using a test application and a test server from a Windows Sever 2019 ISO. It can be done correctly, successfully, and is supported in .NET Framework 4.6 by following the directions I had planned on including as part of this PR:
https://github.com/IonxSolutions/serilog-sinks-syslog/pull/63/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R5

But what I can't figure out is why it doesn't work on AppVeyor's build machines. I hate to have to keep making commits to try and debug it, with the only debugging being logging. Do you know of any other way, or know of any more details about AppVeyor and the Windows Server 2019 image they use?

Kevin-Andrew commented:
P.S. @colin-anderson , I also looked into switching to GitHub Actions, as that was something else we've had on the list:
https://github.com/IonxSolutions/serilog-sinks-syslog/issues/23

But a GitHub Windows Server 2019 "runner" doesn't have .NET Framework 4.6.2 installed on it. See https://github.com/actions/runner-images/blob/main/images/win/Windows2019-Readme.md#net-core-tools

It only has .NET Framework 4.7.2 and 4.8, as well as .NET Core 6. They deprecated .NET Core 3.1 as well.

Compared to what AppVeyor offers on their Windows Server 2019 build machine:
https://www.appveyor.com/docs/windows-images-software/#net-framework

So if switching to GitHub Actions, we'd have to reduce the build targets down to just .NET Framework 4.7.2 and .NET 6.

antibarbie commented:
I guess that somehow the value "ServicePointManager.DisableSystemDefaultTlsVersions" was already loaded before it was overriden with the correct value.  (And a difference with machine settings induce a difference in code paths that prematurely loads "ServicePointManager" ?)

Here the dopcumentation indicates that " _compatibility settings must be the first thing the application does_  " :  https://learn.microsoft.com/en-us/dotnet/framework/migration-guide/mitigation-tls-protocols

Using the compatibility settings in the unit test "app.config" file will prove me right (or wrong...)

colin-anderson commented:
@Kevin-Andrew 

> I hate to have to keep making commits to try and debug it, with the only debugging being logging

Ah, the DevOps Way™, don't you just love it 🤣 

> So if switching to GitHub Actions, we'd have to reduce the build targets down to just .NET Framework 4.7.2 and .NET 6

Dropping .NET Core 2.2 and .NET Core 3.1 should fine, since they're already EOL. We're going to need to keep .NET Framework 4.6.2 tho 😞 

I'm conscious that you've already dug right in and spent a lot of time on finding a solution for this 🙏, so we need to use what you've learned, make a decision and GTM 😃. There's no ideal solution, but I think ignoring the failing test only when run on .NET 4.6.2 on AppVeyor is "good enough", and of course they'll still run locally. You can use an XUnit `FactAttribute` to do this (I've just done this inline, so it might not compile, but I'm sure it's close :)

```c#
public sealed class NotNet462OnAppVeyorFactAttribute : FactAttribute
{
    public NotNet462OnAppVeyorFactAttribute()
    {
        var envValue = Environment.GetEnvironmentVariable("APPVEYOR");
        var isAppVeyor = !String.IsNullOrEmpty(envValue) && envValue.Equals("true", StringComparison.OrdinalIgnoreCase);

#if NET462
       const bool IsNet462 = true;
#else
        const bool IsNet462 = false;
#endif

        // ReSharper disable once ConditionIsAlwaysTrueOrFalse
        if (isAppVeyor && IsNet462)
        {
            this.Skip = "This test is not supported on .NET Framework 4.6.2 when running on AppVeyor";
        }
    }
}
```

This sounds OK?

Kevin-Andrew commented:
@colin-anderson , finally got it working:
![image](https://github.com/IonxSolutions/serilog-sinks-syslog/assets/22459191/3e4b6876-1769-46bd-825e-3bb4fbc82914)

The key was to use an xUnit `AssemblyFixture`, [as mentioned here](https://stackoverflow.com/a/53143426/8169136), to get the  code for the `SystemDefaultTlsVersions` setting to run before any tests run and cause the `ServicePointManager` to load its configuration (which it only does once).

@antibarbie , I had tried the "app.config" approach, mentioned in my [comment here](https://github.com/IonxSolutions/serilog-sinks-syslog/pull/63#issuecomment-1606258670). But "app.configs" for DLLs don't get automatically loaded. So I also tried an approach [mentioned here](https://stackoverflow.com/a/74357802/8169136) for unit tests. But that didn't work either.

However, the general principal of ensuring the setting is configured before accessing the `ServicePointManager` (even though we don't explicitly access the `ServicePointManager` class ourselves), was the key.

@colin-anderson , I'll let you take one more look at this, and then either you can merge it (please go ahead and Squash merge it) and make a release, or let me know and I can merge the PR, but you'd still need to be the one to create the release.

antibarbie commented:
@Kevin-Andrew Well done Kevin !   To my experience, in NUnit, the app.config (testproject.dll.config) are actually taken into account correctly (at least as far as binding redirects are concerned) ; hence my suggestion. That's probably not the case with XUnit. Nevertheless, settings the correct value in the fixture init feels better IMHO.

colin-anderson commented:
@Kevin-Andrew damn, you were determined to figure this out! Many thanks for this! 😄
